module/extension/Director/src/bits/challenge_priority.cpp

```diff
@@ -0,0 +1,104 @@
+ * The NUbots Codebase is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The NUbots Codebase is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2021 NUbots <nubots@nubots.net>
+ */
+
+#include <algorithm>
+#include <tuple>
+#include <vector>
+
+#include "Director.hpp"
+
+#include "extension/Behaviour.hpp"
+
+namespace module::extension {
+
+    using ::extension::behaviour::commands::DirectorTask;
+
+    namespace {
+        struct TaskPriority {
+            TaskPriority(const uint64_t& id_, const int& priority_, const bool& optional_)
+                : id(id_), priority(priority_), optional(optional_) {}
+
+            uint64_t id;
+            int priority;
+            bool optional;
+        };
+    }  // namespace
+
+    bool Director::challenge_priority(const std::shared_ptr<const DirectorTask>& incumbent,
+                                      const std::shared_ptr<const DirectorTask>& challenger) {
+
+        // Compare to yourself and you get that you shouldn't change tasks
+        if (incumbent->requester_id == challenger->requester_id) {
+            return false;
+        }
+
+        // Function to get the priorities of the ancestors of this task
+        auto get_ancestor_priorities = [this](const std::shared_ptr<const DirectorTask>& task) {
+            // We are our first anscestor
+            std::vector<TaskPriority> ancestors;
+            ancestors.emplace_back(task->requester_id, task->priority, task->optional);
+
+            // Loop up through the providers until we reach a point where a task was emitted by a non provider
+            for (auto t = task; providers.count(t->requester_id) != 0;) {
+
+                // Get the provider that emitted this task, and from that the provider group
+                auto provider = providers[t->requester_id];
+                auto& group   = groups[provider->type];
+
+                // If there is no active task something has gone wrong with the algorithm
+                if (group.active_task == nullptr) {
+                    throw std::runtime_error("Task has broken parentage");
+                }
+
+                // Add this new parent we found and set t so we can loop up further
+                t = group.active_task;
+                ancestors.emplace_back(t->requester_id, t->priority, t->optional);
+            }
+            return ancestors;
+        };
+
+        // Convenience references for the two to make the code shorter
```

Comment:
These variables are only used once in the following line it seems unnecessary to declare them explicitly unless you plan to extend this algorithm further

Reply:
Fine I'll remove them!

Reply:
I just meant those two lines but getting rid of a and b entirely works too I guess 😆